File: datastructures/src/main/java/com/javaground/datastructures/stack/fixed/FixedSizedStack.java
package com.javaground.datastructures.stack.fixed;

import com.javaground.datastructures.stack.Stack;
import org.jetbrains.annotations.NotNull;

import java.util.Iterator;

/**
 * A fixed-size stack implementation that extends the abstract Stack class.
 * This implementation uses a pre-allocated array to store elements, ensuring
 * the stack has a set maximum capacity defined at the time of its creation.
 *
 * <p>Improvements:
 * <ul>
 *   <li>Correct size tracking using a {@code currentSize} counter</li>
 *   <li>Bounds checks for overflow/underflow in {@code push}, {@code pop}, and {@code peek}</li>
 *   <li>Return top element correctly and clear popped slots to help GC</li>
 *   <li>Accurate implementations of {@code size()} and {@code isEmpty()}</li>
 * </ul>
 *
 * @param <T> the type of elements stored in the stack
 */
public class FixedSizedStack<T> extends Stack<T> {
    private int currentSize = 0;
    private final T[] items;

    @SuppressWarnings("unchecked")
    public FixedSizedStack(int maxsize) {
        if (maxsize <= 0) {
            throw new IllegalArgumentException("maxsize must be > 0, was " + maxsize);
        }
        this.items = (T[]) new Object[maxsize];
    }

    @Override
    public void push(T item) {
        if (currentSize == items.length) {
            throw new IllegalStateException("Stack overflow: capacity=" + items.length);
        }
        items[currentSize++] = item;
    }

    @Override
    public T pop() {
        if (currentSize == 0) {
            throw new IllegalStateException("Stack underflow: empty stack");
        }
        T value = items[--currentSize];
        // help GC by preventing loitering of unused items by setting them to null
        items[currentSize] = null;
        return value;
    }

    @Override
    public T peek() {
        if (currentSize == 0) {
            throw new IllegalStateException("Empty stack");
        }
        return items[currentSize - 1];
    }

    @Override
    public Integer size() {
        return currentSize;
    }

    @Override
    public Boolean isEmpty() {
        return currentSize == 0;
    }

    @Override
    public @NotNull Iterator<T> iterator() {
        return new FixedSizedStackIterator();
    }

    public class FixedSizedStackIterator implements Iterator<T> {
        @Override
        public boolean hasNext() {
            return currentSize > 0;
        }

        @Override
        public T next() {
            return items[--currentSize];
        }
    }
}

File: datastructures/src/main/kotlin/com/kotlinground/datastructures/stack/fixed/FixedSizedStack.kt
package com.kotlinground.datastructures.stack.fixed

import com.kotlinground.datastructures.stack.Stack

/**
 * A stack implementation with a fixed maximum size. This stack prevents dynamically resizing the storage
 * and enforces a strict limit on the number of elements it can hold.
 *
 * @param T the type of elements held in this stack.
 * @param maxSize the maximum number of elements the stack can hold (must be > 0).
 */
class FixedSizedStack<T : Comparable<T>>(maxSize: Int) : Stack<T>() {
    private val items: Array<Any?>
    private var currentSize = 0

    init {
        require(maxSize > 0) { "maxSize must be > 0, was $maxSize" }
        items = arrayOfNulls(maxSize)
    }

    override fun push(item: T) {
        if (currentSize == items.size) {
            throw IllegalStateException("Stack overflow: capacity=${items.size}")
        }
        items[currentSize++] = item
    }

    override fun pop(): T {
        if (currentSize == 0) {
            throw IllegalStateException("Stack underflow: empty stack")
        }
        val idx = --currentSize
        @Suppress("UNCHECKED_CAST")
        val value = items[idx] as T
        // help GC by preventing loitering of unused items by setting them to null
        items[idx] = null
        return value
    }

    override fun peek(): T {
        if (currentSize == 0) {
            throw IllegalStateException("Empty stack")
        }
        @Suppress("UNCHECKED_CAST")
        return items[currentSize - 1] as T
    }

    override val size: Int
        get() = currentSize

    override val isEmpty: Boolean
        get() = currentSize == 0

    override fun next(): T {
        @Suppress("UNCHECKED_CAST")
        return items[--currentSize] as T
    }

    override fun hasNext(): Boolean {
        return currentSize > 0
    }
}
